### ISSUE TITLE
check-networking-bond TestNetworking.testBondingMain flake: waiting for non-existing "eth1"

### ISSUE THREAD

**Issue by martinpitt** [cockpit-project/cockpit#10265]
We often see this test case [fail](https://fedorapeople.org/groups/cockpit/logs/pull-10260-20181010-071313-80c39d2c-verify-rhel-7-6-distropkg/log.html#51) like this:
```
  File "test/verify/check-networking-bond", line 216, in testBondingMain
    b.wait_present("#networking-interfaces tr[data-interface='%s']" % iface)
testlib.Error: timeout
```

The screenshot shows that there is indeed no eth1 in the UI:

![screeshot](https://fedorapeople.org/groups/cockpit/logs/pull-10260-20181010-071313-80c39d2c-verify-rhel-7-6-distropkg/TestNetworking-testBondingMain-rhel-7-6-distropkg-127.0.0.2-2301-FAIL.png)

This bug reproduces locally for me in about 1 out of 3 cases (with some background load, like running another test in parallel).

**Comment by martinpitt** [comment]
When this happens, I can log into Cockpit and still see no eth1, so it's not just a race/missing signal in the UI.

`ip -a` for these two interfaces shows that eth1 exists, is up, but has no IP address.
```
2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast state UP group default qlen 1000
    link/ether 52:54:00:12:34:56 brd ff:ff:ff:ff:ff:ff
    inet 172.27.0.15/24 brd 172.27.0.255 scope global noprefixroute dynamic eth0
       valid_lft 85973sec preferred_lft 85973sec
    inet6 fec0::5054:ff:fe12:3456/64 scope site mngtmpaddr dynamic 
       valid_lft 85978sec preferred_lft 13978sec
    inet6 fe80::5054:ff:fe12:3456/64 scope link 
       valid_lft forever preferred_lft forever
3: eth1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast state UP group default qlen 1000
    link/ether 52:54:01:00:00:01 brd ff:ff:ff:ff:ff:ff
```

```
# nmcli d show eth1
GENERAL.DEVICE:                         eth1
GENERAL.TYPE:                           ethernet
GENERAL.HWADDR:                         52:54:01:00:00:01
GENERAL.MTU:                            1500
GENERAL.STATE:                          100 (connected)
GENERAL.CONNECTION:                     static-52:54:01
GENERAL.CON-PATH:                       /org/freedesktop/NetworkManager/ActiveConnection/7
WIRED-PROPERTIES.CARRIER:               on
IP4.GATEWAY:                            --
IP6.GATEWAY:                            --
IP6.ROUTE[1]:                           dst = ff00::/8, nh = ::, mt = 256, table=255
```

There are no bond or other devices any more (this happens after deleting the bond from the UI), only lo and virbr0* (which seem unrelated).

When I click on "Add Bond", I do see eth1 being available in the "members" list.

When the test case succeeds, the status is like this instead:
```
3: eth1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast state UP group default qlen 1000
    link/ether 52:54:01:00:00:01 brd ff:ff:ff:ff:ff:ff
    inet 10.111.118.223/20 brd 10.111.127.255 scope global noprefixroute dynamic eth1
       valid_lft 3584sec preferred_lft 3584sec
    inet6 fe80::4942:7c67:7366:4677/64 scope link noprefixroute 
       valid_lft forever preferred_lft forever

# nmcli d show eth1
GENERAL.DEVICE:                         eth1
GENERAL.TYPE:                           ethernet
GENERAL.HWADDR:                         52:54:01:00:00:01
GENERAL.MTU:                            1500
GENERAL.STATE:                          100 (connected)
GENERAL.CONNECTION:                     static-52:54:01
GENERAL.CON-PATH:                       /org/freedesktop/NetworkManager/ActiveConnection/8
WIRED-PROPERTIES.CARRIER:               on
IP4.ADDRESS[1]:                         10.111.118.223/20
IP4.GATEWAY:                            10.111.113.2
IP4.ROUTE[1]:                           dst = 0.0.0.0/0, nh = 10.111.113.2, mt = 101
IP4.ROUTE[2]:                           dst = 10.111.112.0/20, nh = 0.0.0.0, mt = 101
IP4.DNS[1]:                             10.111.113.2
IP4.DOMAIN[1]:                          cockpit.lan
IP6.ADDRESS[1]:                         fe80::4942:7c67:7366:4677/64
IP6.GATEWAY:                            --
IP6.ROUTE[1]:                           dst = fe80::/64, nh = ::, mt = 101
IP6.ROUTE[2]:                           dst = ff00::/8, nh = ::, mt = 256, table=255
```

@mvollmer, does that ring a bell?

**Comment by mvollmer** [comment]
So eth1 is first a slave in a bond, then we delete that bond, and we expect eth1 to appear again as a non-slave device.

Maybe eth1 is still configured as a slave of a now non-existing bond.  Slaves are not shown on the overview page.  (But I should check what happens with slaves of non-existing devices. We should let the user reclaim them somehow.)

I'll try to reproduce this as well, but if you have time, could you show `nmcli con show static-52:54:01`?  Does anything change after `nmcli con reload`?

**Comment by martinpitt** [comment]
```
# nmcli con show static-52:54:01
connection.id:                          static-52:54:01
connection.uuid:                        46bdafb1-16fc-41d5-859c-58a77eac9fb8
connection.stable-id:                   --
connection.type:                        802-3-ethernet
connection.interface-name:              eth1
connection.autoconnect:                 yes
connection.autoconnect-priority:        0
connection.autoconnect-retries:         -1 (default)
connection.auth-retries:                -1
connection.timestamp:                   1539176049
connection.read-only:                   no
connection.permissions:                 --
connection.zone:                        --
connection.master:                      --
connection.slave-type:                  --
connection.autoconnect-slaves:          -1 (default)
connection.secondaries:                 --
connection.gateway-ping-timeout:        0
connection.metered:                     unknown
connection.lldp:                        default
802-3-ethernet.port:                    --
802-3-ethernet.speed:                   0
802-3-ethernet.duplex:                  --
802-3-ethernet.auto-negotiate:          no
802-3-ethernet.mac-address:             --
802-3-ethernet.cloned-mac-address:      --
802-3-ethernet.generate-mac-address-mask:--
802-3-ethernet.mac-address-blacklist:   --
802-3-ethernet.mtu:                     auto
802-3-ethernet.s390-subchannels:        --
802-3-ethernet.s390-nettype:            --
802-3-ethernet.s390-options:            --
802-3-ethernet.wake-on-lan:             default
802-3-ethernet.wake-on-lan-password:    --
ipv4.method:                            auto
ipv4.dns:                               --
ipv4.dns-search:                        --
ipv4.dns-options:                       ""
ipv4.dns-priority:                      0
ipv4.addresses:                         --
ipv4.gateway:                           --
ipv4.routes:                            --
ipv4.route-metric:                      -1
ipv4.route-table:                       0 (unspec)
ipv4.ignore-auto-routes:                no
ipv4.ignore-auto-dns:                   no
ipv4.dhcp-client-id:                    --
ipv4.dhcp-timeout:                      0 (default)
ipv4.dhcp-send-hostname:                yes
ipv4.dhcp-hostname:                     --
ipv4.dhcp-fqdn:                         --
ipv4.never-default:                     no
ipv4.may-fail:                          yes
ipv4.dad-timeout:                       -1 (default)
ipv6.method:                            auto
ipv6.dns:                               --
ipv6.dns-search:                        --
ipv6.dns-options:                       ""
ipv6.dns-priority:                      0
ipv6.addresses:                         --
ipv6.gateway:                           --
ipv6.routes:                            --
ipv6.route-metric:                      -1
ipv6.route-table:                       0 (unspec)
ipv6.ignore-auto-routes:                no
ipv6.ignore-auto-dns:                   no
ipv6.never-default:                     no
ipv6.may-fail:                          yes
ipv6.ip6-privacy:                       -1 (unknown)
ipv6.addr-gen-mode:                     stable-privacy
ipv6.dhcp-send-hostname:                yes
ipv6.dhcp-hostname:                     --
ipv6.token:                             --
proxy.method:                           none
proxy.browser-only:                     no
proxy.pac-url:                          --
proxy.pac-script:                       --
GENERAL.NAME:                           static-52:54:01
[root@localhost ~]# nmcli con show static-52:54:01|cat
connection.id:                          static-52:54:01
connection.uuid:                        46bdafb1-16fc-41d5-859c-58a77eac9fb8
connection.stable-id:                   --
connection.type:                        802-3-ethernet
connection.interface-name:              eth1
connection.autoconnect:                 yes
connection.autoconnect-priority:        0
connection.autoconnect-retries:         -1 (default)
connection.auth-retries:                -1
connection.timestamp:                   1539176049
connection.read-only:                   no
connection.permissions:                 --
connection.zone:                        --
connection.master:                      --
connection.slave-type:                  --
connection.autoconnect-slaves:          -1 (default)
connection.secondaries:                 --
connection.gateway-ping-timeout:        0
connection.metered:                     unknown
connection.lldp:                        default
802-3-ethernet.port:                    --
802-3-ethernet.speed:                   0
802-3-ethernet.duplex:                  --
802-3-ethernet.auto-negotiate:          no
802-3-ethernet.mac-address:             --
802-3-ethernet.cloned-mac-address:      --
802-3-ethernet.generate-mac-address-mask:--
802-3-ethernet.mac-address-blacklist:   --
802-3-ethernet.mtu:                     auto
802-3-ethernet.s390-subchannels:        --
802-3-ethernet.s390-nettype:            --
802-3-ethernet.s390-options:            --
802-3-ethernet.wake-on-lan:             default
802-3-ethernet.wake-on-lan-password:    --
ipv4.method:                            auto
ipv4.dns:                               --
ipv4.dns-search:                        --
ipv4.dns-options:                       ""
ipv4.dns-priority:                      0
ipv4.addresses:                         --
ipv4.gateway:                           --
ipv4.routes:                            --
ipv4.route-metric:                      -1
ipv4.route-table:                       0 (unspec)
ipv4.ignore-auto-routes:                no
ipv4.ignore-auto-dns:                   no
ipv4.dhcp-client-id:                    --
ipv4.dhcp-timeout:                      0 (default)
ipv4.dhcp-send-hostname:                yes
ipv4.dhcp-hostname:                     --
ipv4.dhcp-fqdn:                         --
ipv4.never-default:                     no
ipv4.may-fail:                          yes
ipv4.dad-timeout:                       -1 (default)
ipv6.method:                            auto
ipv6.dns:                               --
ipv6.dns-search:                        --
ipv6.dns-options:                       ""
ipv6.dns-priority:                      0
ipv6.addresses:                         --
ipv6.gateway:                           --
ipv6.routes:                            --
ipv6.route-metric:                      -1
ipv6.route-table:                       0 (unspec)
ipv6.ignore-auto-routes:                no
ipv6.ignore-auto-dns:                   no
ipv6.never-default:                     no
ipv6.may-fail:                          yes
ipv6.ip6-privacy:                       -1 (unknown)
ipv6.addr-gen-mode:                     stable-privacy
ipv6.dhcp-send-hostname:                yes
ipv6.dhcp-hostname:                     --
ipv6.token:                             --
proxy.method:                           none
proxy.browser-only:                     no
proxy.pac-url:                          --
proxy.pac-script:                       --
GENERAL.NAME:                           static-52:54:01
GENERAL.UUID:                           46bdafb1-16fc-41d5-859c-58a77eac9fb8
GENERAL.DEVICES:                        eth1
GENERAL.STATE:                          activated
GENERAL.DEFAULT:                        no
GENERAL.DEFAULT6:                       no
GENERAL.SPEC-OBJECT:                    --
GENERAL.VPN:                            no
GENERAL.DBUS-PATH:                      /org/freedesktop/NetworkManager/ActiveConnection/7
GENERAL.CON-PATH:                       /org/freedesktop/NetworkManager/Settings/6
GENERAL.ZONE:                           --
GENERAL.MASTER-PATH:                    --
IP4.GATEWAY:                            --
IP6.GATEWAY:                            --
IP6.ROUTE[1]:                           dst = ff00::/8, nh = ::, mt = 256, table=255
```

After `nmcli con reload` there is zero change in any of `ip -a`, `nmcli d`, and `nmcli con show static-52:54:01`. I keep that VM running for now, if you need any other info.

**Comment by mvollmer** [comment]
When eth1 is missing in the UI, it still has a ActiveConnection with a Master.

Introspecting the D-Bus API shows that master, but `nmcli con show` does not.  The Master property in the API points to a non-existing D-Bus object.

Cockpit makes the API calls for deleting the bond and freeing its slaves in parallel, so I guess we are asking for triggering a race condition in NM.  The fix might be to do these calls strictly sequential.
Also, we can make Cockpit more robust and treat pointers to non-existing objects as null.

```
# gdbus introspect -y -d org.freedesktop.NetworkManager -o /org/freedesktop/NetworkManager/ActiveConnection/7
node /org/freedesktop/NetworkManager/ActiveConnection/7 {
  interface org.freedesktop.DBus.Properties {
    methods:
      Get(in  s interface_name,
          in  s property_name,
          out v value);
      GetAll(in  s interface_name,
             out a{sv} properties);
      Set(in  s interface_name,
          in  s property_name,
          in  v value);
    signals:
      PropertiesChanged(s interface_name,
                        a{sv} changed_properties,
                        as invalidated_properties);
    properties:
  };
  interface org.freedesktop.DBus.Introspectable {
    methods:
      Introspect(out s xml_data);
    signals:
    properties:
  };
  interface org.freedesktop.DBus.Peer {
    methods:
      Ping();
      GetMachineId(out s machine_uuid);
    signals:
    properties:
  };
  interface org.freedesktop.NetworkManager.Connection.Active {
    methods:
    signals:
      StateChanged(u state,
                   u reason);
      PropertiesChanged(a{sv} properties);
    properties:
      readonly o Connection = '/org/freedesktop/NetworkManager/Settings/6';
      readonly o SpecificObject = '/';
      readonly s Id = 'static-52:54:01';
      readonly s Uuid = 'e8a96c32-a12d-4727-9018-a6dfc21c2664';
      readonly s Type = '802-3-ethernet';
      readonly ao Devices = ['/org/freedesktop/NetworkManager/Devices/3'];
      readonly u State = 2;
      readonly u StateFlags = 30;
      readonly b Default = false;
      readonly o Ip4Config = '/org/freedesktop/NetworkManager/IP4Config/8';
      readonly o Dhcp4Config = '/';
      readonly b Default6 = false;
      readonly o Ip6Config = '/org/freedesktop/NetworkManager/IP6Config/8';
      readonly o Dhcp6Config = '/';
      readonly b Vpn = false;
      readonly o Master = '/org/freedesktop/NetworkManager/Devices/6';
  };
};

# nmcli con show /org/freedesktop/NetworkManager/ActiveConnection/7 | grep MASTER
GENERAL.MASTER-PATH:                    --

# gdbus introspect -y -d org.freedesktop.NetworkManager -o / -r | grep "node /org/freedesktop/NetworkManager/Devices"
        node /org/freedesktop/NetworkManager/Devices {
          node /org/freedesktop/NetworkManager/Devices/3 {
          node /org/freedesktop/NetworkManager/Devices/2 {
          node /org/freedesktop/NetworkManager/Devices/1 {
          node /org/freedesktop/NetworkManager/Devices/5 {
          node /org/freedesktop/NetworkManager/Devices/4 {
```

**Comment by martinpitt** [comment]
Should have been fixed with PR #10303.
